File: functions/__tests__/applyXp.test.js
const { applyXp } = require('../xp');

describe('applyXp', () => {
  test('950 +50 -> level up', () => {
    const res = applyXp({ xp: 950, level: 1, add: 50 });
    expect(res).toEqual({ xp: 0, level: 2, leveledUp: true });
  });

  test('multiple overflow', () => {
    const res = applyXp({ xp: 1950, level: 1, add: 200 });
    expect(res).toEqual({ xp: 150, level: 3, leveledUp: true });
  });

  test('at max level 30', () => {
    const res = applyXp({ xp: 900, level: 30, add: 200 });
    expect(res).toEqual({ xp: 0, level: 30, leveledUp: false });
  });
});

File: test/apply_xp_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tapem/features/xp/domain/apply_xp.dart';

void main() {
  test('950 +50 -> level up', () {
    final res = applyXp(xp: 950, level: 1, add: 50);
    expect(res.xp, 0);
    expect(res.level, 2);
    expect(res.leveledUp, true);
  });

  test('multiple overflow', () {
    final res = applyXp(xp: 1950, level: 1, add: 200);
    expect(res.xp, 150);
    expect(res.level, 3);
    expect(res.leveledUp, true);
  });

  test('at max level', () {
    final res = applyXp(xp: 900, level: 30, add: 200);
    expect(res.xp, 0);
    expect(res.level, 30);
    expect(res.leveledUp, false);
  });
}
